### PR TITLE
[rtsan] Introduce function-name-matches suppression

### PR DESCRIPTION
**Pull request by cjappl**
Introduces a new types of suppression:

1. function-name-matches - allows users to disable `malloc`, `free`, `pthread_mutex_lock` or similar. This could be helpful if a user thinks these are real-time safe on their OS. Also allows disabling of any function marked [[blocking]].

This is useful as a **more performant "early outs" compared to the `call-stack-contains` suppression**. `call-stack-contains` is inherently VERY costly, needing to inspect every frame of every stack for a matching string. This new suppression has an early out before we unwind the stack.

### CODE REVIEW

File: compiler-rt/test/rtsan/stack_suppressions.cpp [hunk 1]
```diff
@@ -1,4 +1,5 @@
 // RUN: %clangxx -fsanitize=realtime %s -o %t
+// RUN: %env_rtsan_opts=halt_on_error=false %run %t 2>&1 | FileCheck %s --check-prefix=CHECK-NOSUPPRESSIONS
```
Comment (cjappl) [inline compiler-rt/test/rtsan/stack_suppressions.cpp:2]: Added this check here to make sure that these would actually cause errors before suppressing them

File: compiler-rt/lib/rtsan/rtsan_checks.inc [hunk 2]
```diff
@@ -17,3 +17,5 @@
 // SummaryKind should be a string literal.
 
 RTSAN_CHECK(CallStackContains, "call-stack-contains")
+RTSAN_CHECK(InterceptedFnName, "intercepted-fn-name")
+RTSAN_CHECK(BlockingFnName, "blocking-fn-name")
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/rtsan_checks.inc:21]: Any feedback on the names?

File: compiler-rt/lib/rtsan/rtsan_suppressions.cpp [hunk 3]
```diff
@@ -92,3 +92,25 @@ bool __rtsan::IsStackTraceSuppressed(const StackTrace &stack) {
...
+
+bool __rtsan::IsInterceptorSuppressed(const char *interceptor_name) {
+  return IsFunctionSuppressed(
+      interceptor_name, ConvertTypeToFlagName(ErrorType::InterceptedFnName));
+}
+
+bool __rtsan::IsBlockingFnSuppressed(const char *blocking_fn_name) {
+  return IsFunctionSuppressed(blocking_fn_name,
+                              ConvertTypeToFlagName(ErrorType::BlockingFnName));
+}
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/rtsan_suppressions.cpp:116]: As you can tell, these are very closely related, sharing the same implementation, and just different names based on if they are blocking or intercepted function.

Is it worthwhile to have two categories for this, or lump them into one? Open to feedback here.

**Review by davidtrevelyan — APPROVED** [review]

**Comment by fmayer** [comment]
High level question: why do we distinguish whether it's intercepted or blocking? Does the user care?

**Comment by cjappl** [comment]
> High level question: why do we distinguish whether it's intercepted or blocking? Does the user care?

Good question. I think the answer is "yes, the user cares".

A lot of this complexity (why we distinguish between intercepted and blocking) is to show the user a different error message:

For malloc, an intercepted call:
```
==51105==ERROR: RealtimeSanitizer: unsafe-library-call
Intercepted call to real-time unsafe function `malloc` in real-time context!
```

For a user defined [[clang::blocking]] call:
```
==51326==ERROR: RealtimeSanitizer: blocking-call
Call to blocking function `my_special_allocation()` in real-time context!
```

I think this is useful because it indicates whether or not this is directly controllable by code outside of glibc. If I get the second, I will know that some dev on my team, or library, marked this call as [[blocking]], and I should perhaps interrogate that source code to see if I agree with the attribute.

To simplify our code a lot, we **could** move to a unified error message, and treat both issues as the same, something like:
```
==51326==ERROR: RealtimeSanitizer: unsafe-call
Call to non-real-time-safe function `my_special_allocation()` in real-time context!
...
==51326==ERROR: RealtimeSanitizer: unsafe-call
Call to non-real-time-safe function `malloc()` in real-time context!
```

But I think that we might lose some important detail in the process. The user may behave differently based on if it is an intercepted function vs a [[blocking]] function.

We have discussed a bit if we want to have some notification of an unbound loop. While the details of this (or if it is even possible) are outside the scope of this comment, again I think it would be nice to be notified specifically of what went wrong, instead of a general error message:

```
==51326==ERROR: RealtimeSanitizer: unbound-loop
Potentially unbound loop called from real-time context!
```


WDYT?

**Comment by fmayer** [comment]
> > High level question: why do we distinguish whether it's intercepted or blocking? Does the user care?
> 
> Good question. I think the answer is "yes, the user cares".
> 
> A lot of this complexity (why we distinguish between intercepted and blocking) is to show the user a different error message:
> 
> For malloc, an intercepted call:
> 
> ```
> ==51105==ERROR: RealtimeSanitizer: unsafe-library-call
> Intercepted call to real-time unsafe function `malloc` in real-time context!
> ```
> 
> For a user defined [[clang::blocking]] call:
> 
> ```
> ==51326==ERROR: RealtimeSanitizer: blocking-call
> Call to blocking function `my_special_allocation()` in real-time context!
> ```
> 
> I think this is useful because it indicates whether or not this is directly controllable by code outside of glibc. If I get the second, I will know that some dev on my team, or library, marked this call as [[blocking]], and I should perhaps interrogate that source code to see if I agree with the attribute.
> 
> To simplify our code a lot, we **could** move to a unified error message, and treat both issues as the same, something like:
> 
> ```
> ==51326==ERROR: RealtimeSanitizer: unsafe-call
> Call to non-real-time-safe function `my_special_allocation()` in real-time context!
> ...
> ==51326==ERROR: RealtimeSanitizer: unsafe-call
> Call to non-real-time-safe function `malloc()` in real-time context!
> ```
> 
> But I think that we might lose some important detail in the process. The user may behave differently based on if it is an intercepted function vs a [[blocking]] function.
> 
> We have discussed a bit if we want to have some notification of an unbound loop. While the details of this (or if it is even possible) are outside the scope of this comment, again I think it would be nice to be notified specifically of what went wrong, instead of a general error message:
> 
> ```
> ==51326==ERROR: RealtimeSanitizer: unbound-loop
> Potentially unbound loop called from real-time context!
> ```
> 
> WDYT?

Right, I meant more in the context of this CL though. If I add a list the ignore list, why do I care if it's blocking or intercepted? I understand it might be useful to note why a function is considered "rt unsafe" in the error mesage.

**Comment by cjappl** [comment]
> Right, I meant more in the context of this CL though. If I add a list the ignore list, why do I care if it's blocking or intercepted? 

Good question. @davidtrevelyan any thoughts? We could possibly change this to `function-name`, which would encapsulate both. I'm not opposed to this.

**Comment by davidtrevelyan** [comment]
> > Right, I meant more in the context of this CL though. If I add a list the ignore list, why do I care if it's blocking or intercepted?
> 
> Good question. @davidtrevelyan any thoughts? We could possibly change this to `function-name`, which would encapsulate both. I'm not opposed to this.

Agreed - I think a single `function-name` suppressor would suffice - nice suggestion! I spent a while trying to conceive of a counter example and nothing came to mind.

File: compiler-rt/lib/rtsan/rtsan_checks.inc [hunk 4]
```diff
@@ -17,3 +17,4 @@
 // SummaryKind should be a string literal.
 
 RTSAN_CHECK(CallStackContains, "call-stack-contains")
+RTSAN_CHECK(FunctionNameIs, "function-name-is")
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/rtsan_checks.inc:20]: Please review the new name @davidtrevelyan and @fmayer  - I chose one with a verb at the end to mirror call-stack-contains.

Other candidates I considered:
* unsafe-function-name-is
* function-name
* function-name-matches

Comment (davidtrevelyan) [reply compiler-rt/lib/rtsan/rtsan_checks.inc:20]: I like `function-name-is` if the intention is only to allow users to provide an exact function name match, character for character. But I think that would be difficult to use - how is the user supposed to reason about namespaces, mangling, parentheses on the end, overloads, etc.? I think at the minute I prefer `function-name-matches` with regex supported.

Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan_checks.inc:20]: Cool that works for me. At least as written, we use the suppression built-in simple regexes (which I think is good, and we should keep it). I'll wait to see what fmayer says.

Comment (davidtrevelyan) [reply compiler-rt/lib/rtsan/rtsan_checks.inc:20]: Sounds good!

Comment (fmayer) [reply compiler-rt/lib/rtsan/rtsan_checks.inc:20]: `function-name-matches` SGTM if it's a regex / glob

**Comment by cjappl** [comment]
I'm going to submit this when tests are green, as I think the three of us who have chatted about this are in good agreement. Happy to do anything else as post-commit

**Review by fmayer — APPROVED** [review]

**Pull request merged**